poh/src/leader_bank_notifier.rs

```diff
@@ -0,0 +1,289 @@
+    StandBy,
+    /// PoH-initiated bank is available.
+    InProgress,
+}
+
+#[derive(Debug, Default)]
+struct SlotAndBankWithStatus {
+    status: Status,
+    slot: Option<Slot>,
+    bank: Weak<Bank>,
+}
+
+impl LeaderBankNotifier {
+    /// Set the status to `InProgress` and notify any waiting threads.
+    /// Panics if the status is not `StandBy` - cannot have multiple
+    /// leader banks in progress.
+    pub fn set_in_progress(&self, bank: &Arc<Bank>) {
+        let mut state = self.state.lock().unwrap();
+        assert_eq!(state.status, Status::StandBy);
+
+        *state = SlotAndBankWithStatus {
+            status: Status::InProgress,
+            slot: Some(bank.slot()),
+            bank: Arc::downgrade(bank),
+        };
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// Set the status to `StandBy` and notify any waiting threads if the status was changed.
+    /// Panics if the stored slot does not match the given slot.
+    pub fn set_completed(&self, slot: Slot) {
+        let mut state = self.state.lock().unwrap();
+        assert_eq!(state.slot, Some(slot));
+
+        // `set_completed` can be called multiple times for the same slot because it
+        // may be called from multiple threads.
+        if matches!(state.status, Status::StandBy) {
+            return;
+        }
+
+        state.status = Status::StandBy;
+        drop(state);
+
+        self.condvar.notify_all();
+    }
+
+    /// If the status is `InProgress`, immediately return a weak reference to the bank.
+    /// Otherwise, wait up to the `timeout` for the status to become `InProgress`.
+    /// If the timeout is reached, the weak reference is unupgradable.
+    pub fn get_or_wait_for_in_progress(&self, timeout: Duration) -> Weak<Bank> {
+        let state = self.state.lock().unwrap();
+        self.get_or_wait_for_in_progress_state(timeout, state)
+            .map(|state| state.bank.clone())
+            .unwrap_or_else(Weak::new)
+    }
+
+    /// Wait for next notification for a completed leader slot.
+    /// Returns `None` if the timeout is reached
+    pub fn wait_for_completed(&self, mut remaining_timeout: Duration) -> Option<Slot> {
+        let state = self.state.lock().unwrap();
+
+        // If currently `StandBy`, need to wait for `InProgress` to begin.
+        let now = Instant::now();
+        let state = self.get_or_wait_for_in_progress_state(remaining_timeout, state);
+
+        let Some(state) = state else { return None; };
+        remaining_timeout = remaining_timeout.saturating_sub(now.elapsed());
+
+        // Wait for `StandBy` to be set.
+        let (state, wait_timeout_result) = self
+            .condvar
+            .wait_timeout_while(state, remaining_timeout, |state| {
+                matches!(state.status, Status::InProgress)
+            })
+            .unwrap();
+
+        if !wait_timeout_result.timed_out() {
+            state.slot
+        } else {
+            None
+        }
```

Comment:
nit: `.filter_map()` :)

Reply:
Similarly here, we want to panic on PoisonErrors, and the `wait_timeout_result` is not actually a `Result` type - copied from `condvar.rs`:
```rust
/// A type indicating whether a timed wait on a condition variable returned
/// due to a time out or not.
///
/// It is returned by the [`wait_timeout`] method.
///
/// [`wait_timeout`]: Condvar::wait_timeout
#[derive(Debug, PartialEq, Eq, Copy, Clone)]
#[stable(feature = "wait_timeout", since = "1.5.0")]
pub struct WaitTimeoutResult(bool);
```

The only way we can inspect the "result" is through `timed_out()`, and bools do not implement an `and_then` like options: https://doc.rust-lang.org/std/option/enum.Option.html#method.and_then

Having an if seems cleaner to me than jumping through the hoop of something like:
```rust
    (!wait_timeout_result).then_some(()).and_then(|| state.slot)
```

Reply:
yeah, my bad; we can't use `.filter_map()` here.

how about this? imo, less clumsy than `then_some(...).and_then(...)`:

```patch
diff --git a/poh/src/leader_bank_notifier.rs b/poh/src/leader_bank_notifier.rs
index 309033ba885..e38d64be344 100644
--- a/poh/src/leader_bank_notifier.rs
+++ b/poh/src/leader_bank_notifier.rs
@@ -96,11 +96,7 @@ impl LeaderBankNotifier {
             })
             .unwrap();
 
-        if !wait_timeout_result.timed_out() {
-            state.slot
-        } else {
-            None
-        }
+        (!wait_timeout_result.timed_out()).then(|| state.slot.expect("some slot whenever completed"))
     }
 
     /// Helper function to get or wait for the `InProgress` status with a given `MutexGuard`.
``` 

or

```patch
@@ -96,11 +96,7 @@ impl LeaderBankNotifier {
             })
             .unwrap();

-        if !wait_timeout_result.timed_out() {
-            state.slot
-        } else {
-            None
-        }
+        (!wait_timeout_result.timed_out()).then(|| state.slot).flatten()
     }

     /// Helper function to get or wait for the `InProgress` status with a given `MutexGuard`.

```

`.flatten()` (and current code) is silently discarding fatal logic error of allowign `state.slot` to be `None` at this point.

Reply:
Yeah, we probably should verify that the slot is `Some`: 974cf7ee59d910de56ec44ff3a8cbe93de26ac8a

I don't think it's a super great check since after the first `set_in_progress` the slot will never be cleared, but considering multiple threads could be waiting for complete it's not somethng that is easily done 😄